### PR TITLE
feat(wasm-gen): distinguish between importing and inserting functions

### LINKED ISSUE #3308: wasm-gen: distinguish between importing and inserting functions
### Problem to Solve

https://github.com/gear-tech/gear/blob/6327b8064bc1015caab55a92cfd6f600924a9808/utils/wasm-gen/src/generator/syscalls/imports.rs#L231-L249

### Possible Solution

It is necessary to somehow distinguish between the case when import is really needed for internal purposes (sys_call_dest = source, import is used in sys-call pattern) and when the number of inserted syscalls should be greater than zero.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by StackOverflowExcept1on**
Resolves #3308

TODO:
- [x] rename `SysCallsInjectionAmounts` to `SysCallsInjectionTypes`, `amount.rs` to `injection_type.rs`

### CODE REVIEW

**Comment by StackOverflowExcept1on** [comment]
@CodiumAI-Agent /review

**Comment by StackOverflowExcept1on** [comment]
@CodiumAI-Agent /review

File: utils/wasm-gen/src/config/syscalls/injection.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            };
+
+            for &sys_call_import in required_imports {
+                self.enable_sys_call(InvocableSysCall::Loose(sys_call_import));
+            }
+        }
+    }
+
+    /// Enables given sys-call.
+    pub fn enable_sys_call(&mut self, name: InvocableSysCall) {
```
Comment (SabaunT) [inline utils/wasm-gen/src/config/syscalls/injection.rs:94]:
```suggestion
    pub(crate) fn enable_sys_call(&mut self, name: InvocableSysCall) {
```
This is very low-level method with a specific usage

File: utils/wasm-gen/src/config/syscalls/injection.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+//! Entities describing possible injection types for each sys-call.
+//!
+//! Types here are used to create [`crate::SysCallsConfig`].
+
+use crate::InvocableSysCall;
+
+use gear_wasm_instrument::syscalls::SysCallName;
+use std::{collections::HashMap, ops::RangeInclusive};
+
+/// Possible injection type.
```
Comment (SabaunT) [inline utils/wasm-gen/src/config/syscalls/injection.rs:28]: I would add more docs/clarifications. For example, variants of `InjectionType` should be decsribed thoroughly as it's the code potentially used by entities of external crates.

File: utils/wasm-gen/src/config/syscalls/injection.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+#[derive(Debug, Clone)]
+pub enum InjectionType {
```
Comment (SabaunT) [inline utils/wasm-gen/src/config/syscalls/injection.rs:30]: `InjectionType` type name actually represents idea properly, but it's unclear if you read the crate for the first time. What about `SysCallInjectionKind`?

File: utils/wasm-gen/src/config/syscalls/injection.rs (hunk 2 above)
Comment (SabaunT) [reply utils/wasm-gen/src/config/syscalls/injection.rs:28]: that's not only about the type, but generally the module.

Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls/injection.rs]: done in d65fe0e3ee3c92bfd27098c1eea86963b79e9697

File: utils/wasm-gen/src/config/syscalls/injection.rs (hunk 3 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls/injection.rs]: Can I keep it as `SysCallInjectionType`? "Kind" will probably lead to renaming of `SysCallsInjectionTypes` 
done in 1217f73872b7ef3adeeec2f8f4f2a966a10b6a92

Comment (SabaunT) [reply utils/wasm-gen/src/config/syscalls/injection.rs]: kinds weren't bad, actually.

File: utils/wasm-gen/src/config/syscalls/injection.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+use gear_wasm_instrument::syscalls::SysCallName;
+use std::{collections::HashMap, ops::RangeInclusive};
+
+/// This enum defines how the sys-call should be injected into wasm module.
+#[derive(Debug, Clone)]
+pub enum SysCallInjectionType {
+    /// Don't modify wasm module at all.
+    None,
+    /// Sys-call import will be injected into import section of wasm module,
+    /// but the `wasm-gen` generators will not call that sys-call.
```
Comment (SabaunT) [inline utils/wasm-gen/src/config/syscalls/injection.rs:36]: can you add please that this variant is needed to serve precise syscalls generation and etc.

**Review by SabaunT — APPROVED** [review]

File: utils/wasm-gen/src/config/syscalls/injection.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            self.set(name, min, max);
+        }
+    }
+}
+
+impl Default for SysCallsInjectionTypes {
+    fn default() -> Self {
+        Self::all_once()
+    }
+}
```
Comment (mertwole) [inline utils/wasm-gen/src/config/syscalls/injection.rs:121]: It's very implicit, and as I see we don't use this `Default` impl anywhere

File: utils/wasm-gen/src/generator/syscalls/imports.rs [hunk 6]
```diff
@@ -57,9 +57,16 @@ pub struct SysCallsImportsGeneratorInstantiator<'a, 'b>(
     ),
 );
 
+/// The set of sys-calls that need to be imported to create precise sys-call.
+#[derive(thiserror::Error, Debug)]
+#[error("The following sys-calls must be imported: {0:?}")]
+pub struct RequiredSysCalls(&'static [SysCallName]);
```
Comment (mertwole) [inline utils/wasm-gen/src/generator/syscalls/imports.rs:63]: Move it below `PreciseSysCallError` please

File: utils/wasm-gen/src/config/syscalls/injection.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            };
+
+            for &sys_call_import in required_imports {
+                self.enable_sys_call(InvocableSysCall::Loose(sys_call_import));
+            }
+        }
+    }
+
+    /// Enables given sys-call.
+    pub(crate) fn enable_sys_call(&mut self, name: InvocableSysCall) {
```
Comment (mertwole) [inline utils/wasm-gen/src/config/syscalls/injection.rs:99]:
```suggestion
    pub(crate) fn enable_sys_call_import(&mut self, name: InvocableSysCall) {
```

**Review by mertwole — APPROVED** [review]

File: utils/wasm-gen/src/generator/syscalls/imports.rs (hunk 6 above)
Comment (SabaunT) [reply utils/wasm-gen/src/generator/syscalls/imports.rs:63]: plain type definition is ok to be on the top

File: utils/wasm-gen/src/config/syscalls/injection.rs (hunk 7 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls/injection.rs]: done in 4a837cefad4e38ad9ac6e002d67e2bfac6e603da

File: utils/wasm-gen/src/config/syscalls/injection.rs (hunk 5 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls/injection.rs:121]: This is not my code. It was previously in `SysCallsInjectionAmounts`

File: utils/wasm-gen/src/config/syscalls/injection.rs (hunk 4 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls/injection.rs:36]: done in 0612f9d084ed1d50f7fdfad3e640c5157c225fd7

**Review by breathx — APPROVED** [review]

**Pull request merged**